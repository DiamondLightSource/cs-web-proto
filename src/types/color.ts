import { createCanvas } from "canvas";
import log from "loglevel";
export class Color {
  private r: number;
  private g: number;
  private b: number;
  private a: number;

  public static WHITE = new Color(255, 255, 255);
  public static BLACK = new Color(0, 0, 0);
  public static RED = new Color(255, 0, 0);
  public static GREEN = new Color(0, 128, 0);
  public static BLUE = new Color(0, 0, 255);
  public static YELLOW = new Color(255, 255, 0);
  public static PURPLE = new Color(127, 0, 127);
  public static PINK = new Color(255, 192, 203);
  public static ORANGE = new Color(255, 165, 0);
  public static TRANSPARENT = new Color(0, 0, 0, 0);

  // Use the browser to compute colours from names.
  public static fromName(cssColorName: string): string {
    let ctx = document.createElement("canvas").getContext("2d");
    if (!ctx) {
      // Fall back to the canvas npm package.
      ctx = createCanvas(100, 100).getContext("2d");
    }
    if (ctx) {
      ctx.fillStyle = cssColorName;
      return ctx.fillStyle;
    } else {
      log.warn(
        `Failed to parse color name ${cssColorName} using canvas element.`
      );
      return "#000000";
    }
  }

  public static parse(cssColor: string): Color {
    let r = 0;
    let g = 0;
    let b = 0;
    if (cssColor.startsWith("#")) {
      let intRep = parseInt(cssColor.slice(1), 16);
      b = Math.floor(intRep % 256);
      intRep /= 256;
      g = Math.floor(intRep % 256);
      intRep /= 256;
      r = Math.floor(intRep);
    } else if (cssColor.startsWith("rgb")) {
      const parts = cssColor.match(/rgb\((.*), (.*), (.*)\)/);
      if (parts !== null) {
        r = parseInt(parts[1]);
        g = parseInt(parts[2]);
        b = parseInt(parts[3]);
      }
    } else if (cssColor === "transparent") {
      return Color.TRANSPARENT;
    } else {
      return this.parse(this.fromName(cssColor));
    }

    return new Color(r, g, b);
  }

  public constructor(r: number, g: number, b: number, a = 255) {
    if (r < 0 || r > 255) {
      throw new Error(`r value ${r} out of range`);
    } else if (g < 0 || g > 255) {
      throw new Error(`g value ${g} out of range`);
    } else if (b < 0 || b > 255) {
      throw new Error(`b value ${b} out of range`);
    }

    this.r = r;
    this.g = g;
    this.b = b;
    this.a = a;
  }

  public rgbaString(): string {
    return `rgba(${this.r}, ${this.g}, ${this.b}, ${this.a})`;
  }
}
