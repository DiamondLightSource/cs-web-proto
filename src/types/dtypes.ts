import { Alarm, ALARM_NONE } from "./alarm";
import { Display, DISPLAY_NONE } from "./display";

export class DTime {
  public datetime: Date;

  public constructor(datetime: Date) {
    this.datetime = datetime;
  }
}

export function dtimeNow(): DTime {
  return new DTime(new Date());
}

interface DTypeValue {
  stringValue?: string;
  doubleValue?: number;
  arrayValue?: number[];
}

export class DType {
  public value: DTypeValue;
  public time: DTime;
  public alarm?: Alarm;
  public display?: Display;

  public constructor(
    value: DTypeValue,
    alarm?: Alarm,
    time?: DTime,
    display?: Display
  ) {
    // TODO check for no value.
    this.value = value;
    this.alarm = alarm;
    this.time = time ?? dtimeNow();
    this.display = display;
  }

  public getStringValue(): string {
    if (this.value.stringValue) {
      return this.value.stringValue;
    } else if (this.value.doubleValue) {
      return this.value.doubleValue?.toString();
    } else if (this.value.arrayValue) {
      return this.value.arrayValue?.toString();
    } else {
      return "";
    }
  }

  public getDoubleValue(): number {
    // TODO what if not defined?
    if (this.value.doubleValue) {
      return this.value.doubleValue;
    } else {
      return NaN;
    }
  }

  public getAlarm(): Alarm {
    return this.alarm ?? ALARM_NONE;
  }
  public getTime(): DTime {
    return this.time;
  }
  public getDisplay(): Display | undefined {
    return this.display ?? DISPLAY_NONE;
  }

  public toString(): string {
    return this.getStringValue();
  }
}

export function dtypeToString(dtype?: DType): string {
  if (dtype) {
    return dtype.getStringValue();
  } else {
    return "";
  }
}

export function valueToDtype(
  value: any,
  alarm = ALARM_NONE,
  time = dtimeNow(),
  display = DISPLAY_NONE
): DType {
  const dvalue: DTypeValue = {};
  if (typeof value === "string") {
    dvalue.stringValue = value;
  } else if (typeof value === "number") {
    dvalue.doubleValue = value;
  } else {
    dvalue.doubleValue = 0;
  }
  return new DType(dvalue, alarm, time, display);
}
