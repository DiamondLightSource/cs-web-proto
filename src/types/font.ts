export enum FontStyle {
  Regular,
  Bold,
  Italic,
  BoldItalic
}

export class Font {
  private size: number;
  private style: FontStyle;
  private typeface: string;
  private name?: string;

  public constructor(
    size: number,
    style?: FontStyle,
    typeface?: string,
    name?: string
  ) {
    this.typeface = typeface ?? "Liberation sans";
    this.style = style ?? FontStyle.Regular;
    this.size = size;
    this.name = name;
  }

  public css(): object {
    const fontWeight =
      this.style === FontStyle.Bold || this.style === FontStyle.BoldItalic
        ? "bold"
        : "normal";
    const fontStyle =
      this.style === FontStyle.Italic || this.style === FontStyle.BoldItalic
        ? "italic"
        : "normal";
    return {
      fontFamily: this.typeface,
      fontSize: `${this.size / 10}rem`,
      fontWeight: fontWeight,
      fontStyle: fontStyle
    };
  }
}
