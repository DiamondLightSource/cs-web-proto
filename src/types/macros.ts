import React from "react";

/* A simple dictionary from key to value. */
export interface MacroMap {
  [key: string]: string;
}

/* The MacroContext, used for providing macros to children of
   container widgets.

   Descendants can update the macros of the nearest provider using
   the updateMacro() function.
*/
export interface MacroContextType {
  macros: MacroMap;
  updateMacro: (key: string, value: string) => void;
}
export const MacroContext = React.createContext<MacroContextType>({
  macros: {},
  updateMacro: (key: string, value: string) => {}
});

function interpolate(
  str: string,
  substitutions: { [substitution: string]: string }
): string {
  const requiredMappings: string[] = [];

  const regexp = /\$[{(](.*?)[})]/g;
  let result = null;
  while ((result = regexp.exec(str))) {
    requiredMappings.push(result[1]);
  }
  const missingMappings = requiredMappings.filter(
    (x: string): boolean => substitutions[x] === undefined
  );

  /* Allow missing macros to go unresolved. */
  for (const missing of missingMappings) {
    substitutions[missing] = "${" + missing + "}";
  }

  return str.replace(/\$[{(](.*?)[})]/g, (x, g): string => substitutions[g]);
}

export function resolveMacros(
  unresolvedPvName: string,
  macroMap: MacroMap
): string {
  const resolved = interpolate(unresolvedPvName, macroMap);
  return resolved;
}
