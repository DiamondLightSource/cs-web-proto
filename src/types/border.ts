import { Color } from "./color";

export enum BorderStyle {
  None,
  Line,
  Dashed,
  Dotted,
  GroupBox
}

const CssBorders: { [key in BorderStyle]: string } = {
  [BorderStyle.None]: "none",
  [BorderStyle.Line]: "solid",
  [BorderStyle.Dashed]: "dashed",
  [BorderStyle.Dotted]: "dotted",
  [BorderStyle.GroupBox]: "ridge"
};

export class Border {
  private style: BorderStyle;
  private color: Color;
  // This will render a border width in pixels. I expect that we will revisit this
  // at some later point, possibly to allow or fix to rems.
  private width: number;

  public constructor(style: BorderStyle, color: Color, width: number) {
    this.style = style;
    this.color = color;
    this.width = width;
  }

  public css(): object {
    return {
      borderStyle: CssBorders[this.style],
      borderWidth: `${this.width}px`,
      borderColor: this.color.rgbaString()
    };
  }
}
