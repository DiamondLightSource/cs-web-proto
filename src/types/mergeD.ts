import { DType } from "./dtypes";

export function mergeDtype(original: DType | undefined, update: DType): DType {
  return new DType(
    update.stringValue ?? original?.stringValue,
    update.doubleValue ?? original?.doubleValue,
    update.arrayValue ?? original?.arrayValue,
    update.alarm ?? original?.alarm,
    update.time ?? original?.time,
    update.display ?? original?.display
  );
}
