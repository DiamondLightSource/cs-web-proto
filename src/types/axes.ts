import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public index: number;
  public autoScale?: boolean;
  public autoScaleThreshold?: number;
  public axisColor?: Color;
  public axisTitle?: string;
  public showGrid?: boolean;
  public gridColor?: Color;
  public dashGridLine?: boolean;
  public timeFormat?: number;
  public scaleFormat?: string;
  public scaleFont?: Font;
  public titleFont?: Font;
  public visible?: boolean;
  public logScale?: boolean;
  public leftBottomSide?: boolean;
  public maximum?: number;
  public minimum?: number;
  public yAxis?: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor(idx: number) {
    this.index = idx ?? 0;
    this.autoScale = this.autoScale ?? true;
    this.autoScaleThreshold = this.autoScaleThreshold ?? 0.95;
    this.axisColor = this.axisColor ?? new Color("rgb(0, 0, 0");
    this.axisTitle = this.axisTitle ?? "";
    this.showGrid = this.showGrid ?? false;
    this.gridColor = this.gridColor ?? new Color("rgb(0, 0, 0");
    this.dashGridLine = this.dashGridLine ?? true;
    this.timeFormat = this.timeFormat ?? 0;
    this.scaleFormat = this.scaleFormat ?? "";
    this.scaleFont = this.scaleFont ?? new Font();
    this.titleFont = this.titleFont ?? new Font(10, FontStyle.Bold);
    this.visible = this.visible ?? true;
    this.logScale = this.logScale ?? false;
    this.leftBottomSide = this.leftBottomSide ?? true;
    this.maximum = this.maximum ?? 0;
    this.minimum = this.minimum ?? 100;
    // If axis has index 1 or above, assume y axis
    let axisBool = false;
    if (this.index > 0) {
      axisBool = true;
    }
    this.yAxis = this.yAxis ?? axisBool;
  }
}

export class Axes {
  public count: number;
  public axisOptions: Axis[];

  public constructor(count: number, axes: Axis[]) {
    if (count !== axes.length) {
      throw new Error(
        `Count ${count} is not equal to number of axes ${axes.length}`
      );
    }
    this.count = count;
    this.axisOptions = axes;
  }
}
