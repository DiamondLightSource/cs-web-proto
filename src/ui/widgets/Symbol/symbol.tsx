import React from "react";

import { useCommonCss, Widget } from "../widget";
import { PVWidgetPropType, PVComponent } from "../widgetProps";
import {
  InferWidgetProps,
  BoolPropOpt,
  StringPropOpt,
  ColorPropOpt,
  BoolProp,
  FloatPropOpt,
  BorderPropOpt,
  StringProp,
  ChoicePropOpt,
  FontPropOpt
} from "../propTypes";
import { registerWidget } from "../register";
import { ImageComponent } from "../Image/image";
import { LabelComponent } from "../Label/label";
import { Color } from "../../../types/color";

const SymbolProps = {
  src: StringProp,
  alt: StringPropOpt,
  backgroundColor: ColorPropOpt,
  showLabel: BoolProp,
  labelPosition: ChoicePropOpt([
    "top",
    "left",
    "center",
    "right",
    "bottom",
    "top left",
    "top right",
    "bottom left",
    "bottom right"
  ]),
  border: BorderPropOpt,
  value: FloatPropOpt,
  rotation: FloatPropOpt,
  flipHorizontal: BoolPropOpt,
  flipVertical: BoolPropOpt,
  visible: BoolPropOpt,
  stretchToFit: BoolPropOpt,
  fitToWidth: BoolPropOpt,
  fitToHeight: BoolPropOpt,
  font: FontPropOpt
};

export type SymbolComponentProps = InferWidgetProps<typeof SymbolProps> &
  PVComponent;

/**
 * This component combines the use of a svg with a label, and is used to replace
 * the MultistateMonitorWidget from CS-Studio
 * @param props
 */
export const SymbolComponent = (props: SymbolComponentProps): JSX.Element => {
  const style = useCommonCss(props as any);

  let alignItems = "center";
  let justifyContent = "center";
  switch (props.labelPosition) {
    case "top":
      alignItems = "start";
      break;
    case "right":
      justifyContent = "end";
      break;
    case "bottom":
      alignItems = "end";
      break;
    case "left":
      justifyContent = "start";
      break;
    case "top right":
      alignItems = "start";
      justifyContent = "end";
      break;
    case "bottom right":
      alignItems = "end";
      justifyContent = "end";
      break;
    case "bottom left":
      alignItems = "end";
      justifyContent = "start";
      break;
    case "top left":
      alignItems = "start";
      justifyContent = "start";
      break;
  }

  return (
    <>
      <ImageComponent {...props} />
      {props.showLabel && (
        <>
          <div
            style={{
              ...style,
              backgroundColor: "transparent",
              position: "absolute",
              height: "100%",
              width: "100%",
              top: 0,
              left: 0,
              display: "flex",
              alignItems,
              justifyContent
            }}
          >
            <div style={{ padding: "5%" }}>
              <LabelComponent
                {...props}
                backgroundColor={Color.TRANSPARENT}
                text={props.value?.getStringValue()}
              ></LabelComponent>
            </div>
          </div>
        </>
      )}
    </>
  );
};

const SymbolWidgetProps = {
  ...SymbolProps,
  ...PVWidgetPropType
};

export const Symbol = (
  props: InferWidgetProps<typeof SymbolWidgetProps>
): JSX.Element => <Widget baseWidget={SymbolComponent} {...props} />;

registerWidget(Symbol, SymbolWidgetProps, "symbol");
