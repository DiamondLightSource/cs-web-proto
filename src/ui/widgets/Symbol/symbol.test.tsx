import React from "react";
import { render, screen } from "@testing-library/react";
import { SymbolComponent } from "./symbol";
import { DType } from "../../../types/dtypes";

const fakeValue = new DType({ stringValue: "Fake value" });

describe("<Symbol />", (): void => {
  test("label is not shown if showLabel is false", (): void => {
    const symbolProps = {
      showLabel: false
    };

    render(<SymbolComponent {...(symbolProps as any)} />);

    expect(screen.queryByText("Fake value")).not.toBeInTheDocument();
  });

  test("label is added", (): void => {
    const symbolProps = {
      showLabel: true,
      value: fakeValue
    };
    render(<SymbolComponent {...(symbolProps as any)} />);

    expect(screen.getByText("Fake value")).toBeInTheDocument();
  });

  test("matches snapshot", (): void => {
    const symbolProps = {
      showLabel: true,
      value: fakeValue
    };
    const { asFragment } = render(
      <SymbolComponent {...(symbolProps as any)} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
