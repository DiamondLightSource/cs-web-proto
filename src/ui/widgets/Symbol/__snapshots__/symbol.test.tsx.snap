// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  style={
    Object {
      "backgroundColor": undefined,
      "color": undefined,
      "height": undefined,
      "outline": undefined,
      "outlineOffset": undefined,
      "overflow": "hidden",
      "textAlign": "left",
      "visibility": undefined,
      "width": undefined,
    }
  }
>
  <img
    src="http://localhost:3000/img/undefined"
    style={
      Object {
        "display": "block",
        "height": undefined,
        "transform": "rotate(0deg) scaleX(1) scaleY(1)",
        "width": undefined,
      }
    }
  />
</div>
`;
